arangod/Scheduler/SocketTask.cpp

```diff
@@ -525,44 +419,213 @@ void SocketTask::asyncReadSome() {
   }
 
   // try to read more bytes
-  if (!_abandoned) {
-    if (!reserveMemory()) {
-      LOG_TOPIC(TRACE, Logger::COMMUNICATION) << "failed to reserve memory";
+  if (_abandoned.load(std::memory_order_acquire)) {
+    return;
+  } else if (!reserveMemory()) {
+    LOG_TOPIC(TRACE, Logger::COMMUNICATION) << "failed to reserve memory";
+    return;
+  }
+
+  auto self = shared_from_this();
+
+  // WARNING: the _readBuffer MUST NOT be changed until the callback
+  // has been called! Otherwise ASIO will get confused and write to
+  // the wrong position.
+
+  TRI_ASSERT(_peer != nullptr);
+  _peer->asyncRead(
+      asio_ns::buffer(_readBuffer.end(), READ_BLOCK_SIZE),
+      [self, this](const asio_ns::error_code& ec, std::size_t transferred) {
+        JobGuard guard(_loop);
+        guard.work();
+
+        if (_abandoned.load(std::memory_order_acquire)) {
+          return;
+        } else if (ec) {
+          LOG_TOPIC(DEBUG, Logger::COMMUNICATION)
+              << "read on stream failed with: " << ec.message();
+          closeStream();
+          return;
+        }
+
+        _loop.scheduler->_nrQueued++;
+        _peer->strand.post([self, this, transferred] {
```

Comment:
I think now that this handler is already called in a strand, because of `_strand.wrap`. This `post` call should be redundant now and can be removed